zoomable-image/coil3/src/main/kotlin/me/saket/telephoto/zoomable/coil3/Coil3ImageSource.kt

```diff
@@ -0,0 +1,246 @@
+import coil3.BitmapImage
+import coil3.ImageLoader
+import coil3.compose.asPainter
+import coil3.decode.DataSource
+import coil3.request.CachePolicy
+import coil3.request.ImageRequest
+import coil3.request.ImageResult
+import coil3.request.Options
+import coil3.request.SuccessResult
+import coil3.request.maxBitmapSize
+import coil3.request.transitionFactory
+import coil3.size.Dimension
+import coil3.size.Precision
+import coil3.size.SizeResolver
+import coil3.transition.CrossfadeTransition
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.withContext
+import me.saket.telephoto.subsamplingimage.ImageBitmapOptions
+import me.saket.telephoto.subsamplingimage.SubSamplingImageSource
+import me.saket.telephoto.zoomable.ZoomableImageSource
+import me.saket.telephoto.zoomable.ZoomableImageSource.ResolveResult
+import me.saket.telephoto.zoomable.coil3.Resolver.ImageSourceCreationResult.EligibleForSubSampling
+import me.saket.telephoto.zoomable.coil3.Resolver.ImageSourceCreationResult.ImageDeletedOnlyFromDiskCache
+import me.saket.telephoto.zoomable.internal.RememberWorker
+import me.saket.telephoto.zoomable.internal.copy
+import java.io.File
+import kotlin.math.roundToInt
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.milliseconds
+import coil3.size.Size as CoilSize
+
+internal class Coil3ImageSource(
+  private val model: Any?,
+  private val imageLoader: ImageLoader,
+) : ZoomableImageSource {
+
+  @Composable
+  override fun resolve(canvasSize: Flow<Size>): ResolveResult {
+    val context = LocalContext.current
+    val resolver = remember(this) {
+      Resolver(
+        request = model as? ImageRequest
+          ?: ImageRequest.Builder(context)
+            .data(model)
+            .build(),
+        imageLoader = imageLoader,
+        sizeResolver = { canvasSize.first().toCoilSize() }
+      )
+    }
+    return resolver.resolved
+  }
+
+  private fun Size.toCoilSize() = CoilSize(
+    width = if (width.isFinite()) Dimension(width.roundToInt()) else Dimension.Undefined,
+    height = if (height.isFinite()) Dimension(height.roundToInt()) else Dimension.Undefined
+  )
+}
+
+internal class Resolver(
+  internal val request: ImageRequest,
+  internal val imageLoader: ImageLoader,
+  private val sizeResolver: SizeResolver,
+) : RememberWorker() {
+
+  internal var resolved: ResolveResult by mutableStateOf(
+    ResolveResult(delegate = null)
+  )
+
+  override suspend fun work() {
+    work(skipMemoryCache = false)
+  }
+
+  private suspend fun work(skipMemoryCache: Boolean) {
+    val imageLoader: ImageLoader = imageLoader
+      .newBuilder()
+      // Ignore "no-store" http headers if they're present and always cache images to disk. Otherwise,
+      // telephoto will be unable to sub-sample large images directly from coil's memory cache.
+      //.respectCacheHeaders(false) // todo: how do I do this with coil3?
```

Comment:
@colinrtwhite Since there's no easy way to override this in coil3, I'm reminded about [this issue](https://github.com/coil-kt/coil/issues/1353) about coil offering an API for downloading images. Can I convince you to reconsider it?

While coil has a very nice API for accessing images from its disk cache, there are some nuances related to cache policies and network cache strategies that telephoto (and probably other apps) need to keep in mind. With coil3, there is a new possibility that telephoto might fail to enable sub-sampling for large images because `CacheControlCacheStrategy` was used.

Reply:
@saket What would be your ideal API surface? I think we might be able to make `CacheControlStrategy` configurable per-request so you can set `request.cacheControlStrategy(CacheControlStrategy.DEFAULT)` vs. only when setting up the `ImageLoader`, but I'm curious why Telefoto needs to force always caching. If the user provides an `ImageLoader` with a custom caching strategy wouldn't it expect to follow that strategy?

I think I understand the want for an "always download" request param - the `CacheControlStrategy` should be the source of truth for whether or not to cache for all of Coil 3 (aside from I/O edge cases or setting `diskCache(null)`).

Reply:
Hmm I see we don't have a great way to set this after using `imageLoader.newBuilder`, though. I'll think about this a bit and see what we can do.

Reply:
> I'm curious why Telefoto needs to force always caching. If the user provides an ImageLoader with a custom caching strategy wouldn't it expect to follow that strategy

This is because sub-sampling of images requires reading them from the disk. You're right that this sounds weird that telephoto wants to override the cache strategy, but I'm imagining developers being confused why their app is crashing with an OOM without realizing that their app is disabling caching in a remote part of the codebase.